presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java

```diff
@@ -0,0 +1,246 @@
+        {
+            return cpuCost(getStats(node.getSource()).getOutputSizeInBytes());
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitProject(ProjectNode node, Void context)
+        {
+            return cpuCost(getStats(node).getOutputSizeInBytes());
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitAggregation(AggregationNode node, Void context)
+        {
+            PlanNodeStatsEstimate aggregationStats = getStats(node);
+            PlanNodeStatsEstimate sourceStats = getStats(node.getSource());
+            return PlanNodeCostEstimate.builder()
+                    .setCpuCost(sourceStats.getOutputSizeInBytes())
+                    .setMemoryCost(aggregationStats.getOutputSizeInBytes())
+                    .setNetworkCost(0)
+                    .build();
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitJoin(JoinNode node, Void context)
+        {
+            return calculateJoinCost(
+                    node,
+                    node.getLeft(),
+                    node.getRight(),
+                    node.getDistributionType().orElse(JoinNode.DistributionType.PARTITIONED).equals(JoinNode.DistributionType.REPLICATED));
+        }
+
+        private PlanNodeCostEstimate calculateJoinCost(PlanNode join, PlanNode probe, PlanNode build, boolean replicated)
+        {
+            int numberOfNodesMultiplier = replicated ? numberOfNodes : 1;
+
+            PlanNodeStatsEstimate probeStats = getStats(probe);
+            PlanNodeStatsEstimate buildStats = getStats(build);
+            PlanNodeStatsEstimate outputStats = getStats(join);
+
+            double cpuCost = probeStats.getOutputSizeInBytes() +
+                    buildStats.getOutputSizeInBytes() * numberOfNodesMultiplier +
+                    outputStats.getOutputSizeInBytes();
+
+            if (replicated) {
+                // add the cost of a local repartitioning of build side copies
+                // cost of the repartitioning of a single data copy has been already added in calculateExchangeCost
+                cpuCost += buildStats.getOutputSizeInBytes() * (numberOfNodesMultiplier - 1);
+            }
+
+            double memoryCost = buildStats.getOutputSizeInBytes() * numberOfNodesMultiplier;
+
+            return PlanNodeCostEstimate.builder()
+                    .setCpuCost(cpuCost)
+                    .setMemoryCost(memoryCost)
+                    .setNetworkCost(0)
+                    .build();
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitExchange(ExchangeNode node, Void context)
+        {
+            return calculateExchangeCost(numberOfNodes, getStats(node), node.getType(), node.getScope());
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitTableScan(TableScanNode node, Void context)
+        {
+            return cpuCost(getStats(node).getOutputSizeInBytes()); // TODO: add network cost, based on input size in bytes?
```

Comment:
Because it's not obvious how to factor this in -- actually, ideally this should be provided by a connector. I slightly changed the TODO message to mention this.